File: src/stores/websocketStore.js
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'

export const useWebSocketStore = defineStore('websocket', () => {
  // Server configuration
  const availableServers = ref([
    { id: 'local', name: 'Local Server', url: 'ws://localhost:8765' },
    { id: 'pi', name: 'Pi Server', url: 'ws://192.168.1.195:8765' }
  ])
  const selectedServerId = ref('local') // Default to local
  
  // Connection state
  const isConnected = ref(false)
  const hasRecentSend = ref(false)
  const hasRecentReceive = ref(false)
  const isConnecting = ref(false)
  
  // Message handling
  const messageHandlers = ref(new Map())
  const messageHistory = ref([])
  const maxHistorySize = 50
  
  // Connection event listeners
  const connectionListeners = ref([])
  
  // WebSocket instance and reconnection
  let websocket = null
  let reconnectAttempts = 0
  const maxReconnectAttempts = 5
  let reconnectInterval = null
  let sendTimeout = null
  let receiveTimeout = null

  // Computed
  const serverUrl = computed(() => {
    const server = availableServers.value.find(s => s.id === selectedServerId.value)
    return server ? server.url : availableServers.value[0].url
  })
  
  const currentServer = computed(() => {
    return availableServers.value.find(s => s.id === selectedServerId.value)
  })
  
  const connectionStatus = computed(() => 
    isConnecting.value ? 'Connecting...' : (isConnected.value ? 'Connected' : 'Disconnected')
  )

  // Server selection
  const selectServer = (serverId) => {
    if (selectedServerId.value !== serverId) {
      console.log(`WebSocket: Switching to server: ${serverId}`)
      selectedServerId.value = serverId
      
      // Reconnect to new server
      if (isConnected.value || isConnecting.value) {
        disconnect()
        setTimeout(() => {
          connect()
        }, 100)
      } else {
        connect()
      }
    }
  }

  // Message handler registration
  const registerHandler = (messageType, handler) => {
    messageHandlers.value.set(messageType, handler)
  }

  const unregisterHandler = (messageType) => {
    messageHandlers.value.delete(messageType)
  }

  // Connection event handling
  const addConnectionListener = (listener) => {
    connectionListeners.value.push(listener)
  }

  const removeConnectionListener = (listener) => {
    const index = connectionListeners.value.indexOf(listener)
    if (index > -1) {
      connectionListeners.value.splice(index, 1)
    }
  }

  const triggerConnectionEvent = () => {
    connectionListeners.value.forEach(listener => {
      try {
        listener()
      } catch (error) {
        console.error('Error in connection listener:', error)
      }
    })
  }

  // Core WebSocket methods
  const connect = () => {
    if (isConnecting.value || isConnected.value) {
      return // Already connecting/connected
    }
    
    try {
      isConnecting.value = true
      console.log(`WebSocket: Connecting to ${serverUrl.value}`)
      websocket = new WebSocket(serverUrl.value)
      
      websocket.onopen = () => {
        isConnected.value = true
        isConnecting.value = false
        reconnectAttempts = 0
        console.log(`WebSocket: Connected to ${serverUrl.value}`)
        
        // Trigger connection event for listeners
        triggerConnectionEvent()
      }
      
      websocket.onmessage = (event) => {
        handleIncomingMessage(event.data)
        showReceiveLight()
      }
      
      websocket.onclose = (event) => {
        isConnected.value = false
        isConnecting.value = false
        console.log(`WebSocket: Disconnected from ${serverUrl.value}`)
        
        // Auto-reconnect if not manually disconnected
        if (event.code !== 1000) {
          scheduleReconnect()
        }
      }
      
      websocket.onerror = (error) => {
        console.error(`WebSocket: Connection error to ${serverUrl.value}`)
        isConnecting.value = false
      }
      
    } catch (error) {
      console.error(`WebSocket: Failed to connect to ${serverUrl.value}`)
      isConnecting.value = false
      scheduleReconnect()
    }
  }

  const disconnect = () => {
    if (websocket) {
      websocket.close(1000) // Normal closure
      websocket = null
    }
    isConnected.value = false
    isConnecting.value = false
    
    // Clear any pending reconnect attempts
    if (reconnectInterval) {
      clearTimeout(reconnectInterval)
      reconnectInterval = null
    }
    reconnectAttempts = 0
  }

  const send = (message) => {
    if (websocket && websocket.readyState === WebSocket.OPEN) {
      const messageStr = JSON.stringify(message)
      websocket.send(messageStr)
      addToHistory(message, 'sent')
      showSendLight()
      return true
    } else {
      console.log('WebSocket: Cannot send - not connected')
      return false
    }
  }

  // Message routing
  const handleIncomingMessage = (data) => {
    try {
      const message = JSON.parse(data)
      addToHistory(message, 'received')
      
      if (message.type && messageHandlers.value.has(message.type)) {
        const handler = messageHandlers.value.get(message.type)
        handler(message)
      }
    } catch (error) {
      console.error('WebSocket: Failed to parse incoming message:', error)
    }
  }

  // Reconnection logic
  const scheduleReconnect = () => {
    if (reconnectAttempts < maxReconnectAttempts) {
      reconnectAttempts++
      const delay = Math.min(1000 * Math.pow(2, reconnectAttempts), 30000)
      
      console.log(`WebSocket: Scheduling reconnect attempt ${reconnectAttempts}/${maxReconnectAttempts} in ${delay}ms`)
      
      reconnectInterval = setTimeout(() => {
        connect()
      }, delay)
    } else {
      console.log('WebSocket: Max reconnect attempts reached')
    }
  }

  const reconnect = () => {
    console.log('WebSocket: Manual reconnect requested')
    disconnect()
    reconnectAttempts = 0
    setTimeout(() => {
      connect()
    }, 100)
  }

  // Helper methods
  const showSendLight = () => {
    hasRecentSend.value = true
    if (sendTimeout) clearTimeout(sendTimeout)
    sendTimeout = setTimeout(() => {
      hasRecentSend.value = false
    }, 500)
  }

  const showReceiveLight = () => {
    hasRecentReceive.value = true
    if (receiveTimeout) clearTimeout(receiveTimeout)
    receiveTimeout = setTimeout(() => {
      hasRecentReceive.value = false
    }, 500)
  }

  const addToHistory = (message, direction) => {
    messageHistory.value.unshift({
      message,
      direction,
      timestamp: Date.now()
    })
    
    if (messageHistory.value.length > maxHistorySize) {
      messageHistory.value = messageHistory.value.slice(0, maxHistorySize)
    }
  }

  const clearHistory = () => {
    messageHistory.value = []
  }

  const getHistory = (limit = 10) => {
    return messageHistory.value.slice(0, limit)
  }

  // Auto-connect on store creation
  connect()

  return {
    // State
    isConnected,
    hasRecentSend,
    hasRecentReceive,
    isConnecting,
    availableServers,
    selectedServerId,
    messageHistory,
    
    // Computed
    serverUrl,
    currentServer,
    connectionStatus,
    
    // Methods
    selectServer,
    registerHandler,
    unregisterHandler,
    addConnectionListener,
    removeConnectionListener,
    connect,
    disconnect,
    send,
    reconnect,
    clearHistory,
    getHistory
  }
}) 

File: hardware_gui/lib/stores/websocket_store.dart
import 'dart:async';
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:flutter/foundation.dart';
import 'package:web_socket_channel/web_socket_channel.dart';
import '../models/display_state.dart';
import '../models/websocket_message.dart';
import '../models/parameter.dart';
import 'parameter_store.dart';

class WebSocketStore extends ChangeNotifier {
  final ParameterStore _parameterStore;
  
  WebSocketChannel? _channel;
  WebSocketConnectionState _connectionState = const WebSocketConnectionState();
  Timer? _reconnectTimer;
  Timer? _pingTimer;
  StreamSubscription? _messageSubscription;
  
  // Server configuration
  static const List<ServerInfo> _availableServers = [
    ServerInfo(
      id: 'localhost',
      name: 'Local',
      url: 'ws://localhost:8765',
      isDefault: true,
    ),
    ServerInfo(
      id: 'pi',
      name: 'Pi Bridge',
      url: 'ws://192.168.1.5:8765',
    ),
  ];
  
  String _selectedServerId = 'localhost';
  int _currentServerIndex = 0;
  
  // Activity tracking
  DateTime? _lastSendTime;
  DateTime? _lastReceiveTime;

  WebSocketStore(this._parameterStore) {
    // Set up parameter store WebSocket callback (like Vue.js broadcastValueUpdate)
    _parameterStore.setSendWebSocketMessage(sendMessage);
  }

  // Getters
  WebSocketConnectionState get connectionState => _connectionState;
  bool get isConnected => _connectionState.isConnected;
  bool get isConnecting => _connectionState.isConnecting;
  bool get isDisconnected => _connectionState.isDisconnected;
  bool get isReconnecting => _connectionState.isReconnecting;
  String get serverUrl => _connectionState.url;
  String get connectionStatus => _connectionState.statusText;
  List<ServerInfo> get availableServers => _availableServers;
  String get selectedServerId => _selectedServerId;
  bool get hasRecentSend => _connectionState.hasRecentSend;
  bool get hasRecentReceive => _connectionState.hasRecentReceive;
  
  ServerInfo get currentServer {
    return _availableServers.firstWhere(
      (s) => s.id == _selectedServerId,
      orElse: () => _availableServers.first,
    );
  }

  // Connection management
  Future<void> connect([String? serverId]) async {
    if (_connectionState.isConnected || _connectionState.isConnecting) return;
    
    if (serverId != null) {
      _selectedServerId = serverId;
    }
    
    final server = currentServer;
    
    _updateConnectionState(
      _connectionState.copyWith(
        status: ConnectionStatus.connecting,
        url: server.url,
        reconnectAttempts: 0,
      ),
    );
    
    try {
      _channel = WebSocketChannel.connect(
        Uri.parse(server.url),
      );
      
      _messageSubscription = _channel!.stream.listen(
        _onMessage,
        onError: _onError,
        onDone: _onDisconnected,
      );
      
      _updateConnectionState(
        _connectionState.copyWith(
          status: ConnectionStatus.connected,
          lastPing: DateTime.now(),
        ),
      );
      
      _startPingTimer();
      await _requestParameterState();
      
      debugPrint('Connected to WebSocket: ${server.url}');
      
    } catch (e) {
      debugPrint('WebSocket connection failed: $e');
      _updateConnectionState(
        _connectionState.copyWith(
          status: ConnectionStatus.disconnected,
          errorMessage: e.toString(),
        ),
      );
      
      _scheduleReconnect();
    }
  }

  Future<void> disconnect() async {
    _stopReconnectTimer();
    _stopPingTimer();
    
    await _messageSubscription?.cancel();
    _messageSubscription = null;
    
    await _channel?.sink.close();
    _channel = null;
    
    _updateConnectionState(
      _connectionState.copyWith(
        status: ConnectionStatus.disconnected,
        errorMessage: null,
      ),
    );
    
    debugPrint('Disconnected from WebSocket');
  }

  Future<void> reconnect() async {
    await disconnect();
    await connect();
  }

  // Server selection
  Future<void> selectServer(String serverId) async {
    if (_selectedServerId == serverId) return;
    
    _selectedServerId = serverId;
    
    if (_connectionState.isConnected) {
      await reconnect();
    }
  }

  // Message handling
  void _onMessage(dynamic message) {
    _lastReceiveTime = DateTime.now();
    _updateConnectionState(
      _connectionState.copyWith(lastReceive: _lastReceiveTime),
    );
    
    try {
      final Map<String, dynamic> data = jsonDecode(message as String);
      debugPrint('Received WebSocket message: ${data['type']}');
      
      final wsMessage = WebSocketMessage.fromJson(data);
      _handleWebSocketMessage(wsMessage);
      
    } catch (e) {
      debugPrint('Error parsing WebSocket message: $e');
      debugPrint('Raw message: $message');
      
      // Try to extract message type for debugging
      try {
        final Map<String, dynamic> data = jsonDecode(message as String);
        debugPrint('Message type: ${data['type']}');
        debugPrint('Message data: ${data['data']}');
      } catch (e2) {
        debugPrint('Could not parse message for debugging: $e2');
      }
    }
  }

  void _handleWebSocketMessage(WebSocketMessage message) {
    switch (message.type) {
      case 'parameter_structure_sync':
        _parameterStore.updateParameterStructure(message as ParameterStructureSync);
        break;
        
      case 'parameter_value_sync':
        // Use new method that handles updates array (with shouldBroadcast = false to prevent loops)
        _parameterStore.updateParametersFromMessage(message as ParameterValueSync);
        break;
        
      case 'parameter_color_sync':
        final colorSync = message as ParameterColorSync;
        // Handle color updates array like Vue.js
        for (final update in colorSync.updates) {
          if (update.rgbColor != null) {
            final color = ParameterColor(
              r: update.rgbColor!['r'] ?? 128,
              g: update.rgbColor!['g'] ?? 128,
              b: update.rgbColor!['b'] ?? 128,
            );
            _parameterStore.updateParameterColor(update.id, color);
          }
        }
        break;
        
      case 'batch_parameter_update':
        _parameterStore.updateParametersFromBatch(message as BatchParameterUpdate);
        break;
        
      case 'led_update':
        // Handle LED update (if needed)
        break;
        
      case 'system':
        _handleSystemMessage(message as SystemMessage);
        break;
        
      default:
        debugPrint('Unknown WebSocket message type: ${message.type}');
    }
  }

  void _handleSystemMessage(SystemMessage message) {
    switch (message.command) {
      case 'pong':
        _updateConnectionState(
          _connectionState.copyWith(lastPing: DateTime.now()),
        );
        break;
        
      case 'error':
        debugPrint('Server error: ${message.message}');
        _updateConnectionState(
          _connectionState.copyWith(errorMessage: message.message),
        );
        break;
        
      default:
        debugPrint('System message: ${message.command} - ${message.message}');
    }
  }

  void _onError(error) {
    debugPrint('WebSocket error: $error');
    _updateConnectionState(
      _connectionState.copyWith(
        status: ConnectionStatus.disconnected,
        errorMessage: error.toString(),
      ),
    );
    
    _scheduleReconnect();
  }

  void _onDisconnected() {
    debugPrint('WebSocket disconnected');
    _updateConnectionState(
      _connectionState.copyWith(
        status: ConnectionStatus.disconnected,
      ),
    );
    
    _scheduleReconnect();
  }

  // Message sending
  Future<void> sendMessage(WebSocketMessage message) async {
    if (!_connectionState.isConnected) return;
    
    try {
      final jsonString = message.toJsonString();
      _channel?.sink.add(jsonString);
      
      _lastSendTime = DateTime.now();
      _updateConnectionState(
        _connectionState.copyWith(lastSend: _lastSendTime),
      );
      
    } catch (e) {
      debugPrint('Error sending WebSocket message: $e');
    }
  }

  // Convenience methods for common messages  
  Future<void> sendParameterUpdate(String id, double value) async {
    final parameter = _parameterStore.getParameterById(id);
    if (parameter == null) return;
    
    final message = ParameterValueSync(
      updates: [
        ParameterValueUpdate(
          id: id,
          value: value,
          text: parameter.text,
          rgbColor: {
            'r': parameter.color.r,
            'g': parameter.color.g,
            'b': parameter.color.b,
          },
        ),
      ],
    );
    
    await sendMessage(message);
  }

  Future<void> sendParameterColorUpdate(String id, ParameterColor color) async {
    final message = ParameterColorSync(
      updates: [
        ParameterColorUpdate(
          id: id,
          rgbColor: {
            'r': color.r,
            'g': color.g,
            'b': color.b,
          },
        ),
      ],
    );
    await sendMessage(message);
  }

  Future<void> _requestParameterState() async {
    final message = RequestParameterState(
      parameterIds: ['all'],
      includeStructure: true,
    );
    await sendMessage(message);
  }

  Future<void> sendPing() async {
    final message = SystemMessage(command: 'ping');
    await sendMessage(message);
  }

  // Reconnection logic
  void _scheduleReconnect() {
    if (_connectionState.reconnectAttempts >= _connectionState.maxReconnectAttempts) {
      debugPrint('Max reconnection attempts reached');
      return;
    }
    
    _stopReconnectTimer();
    
    final delay = _calculateReconnectDelay();
    
    _updateConnectionState(
      _connectionState.copyWith(
        status: ConnectionStatus.reconnecting,
        reconnectAttempts: _connectionState.reconnectAttempts + 1,
        reconnectDelay: delay,
      ),
    );
    
    _reconnectTimer = Timer(Duration(milliseconds: delay), () {
      _tryNextServer();
    });
  }

  void _tryNextServer() {
    _currentServerIndex = (_currentServerIndex + 1) % _availableServers.length;
    final nextServer = _availableServers[_currentServerIndex];
    
    debugPrint('Trying next server: ${nextServer.name}');
    connect(nextServer.id);
  }

  int _calculateReconnectDelay() {
    final baseDelay = 1000; // 1 second
    final maxDelay = 30000; // 30 seconds
    final attempts = _connectionState.reconnectAttempts;
    
    // Exponential backoff: 1s, 2s, 4s, 8s, 16s, 30s (max)
    final delay = (baseDelay * (1 << attempts)).clamp(baseDelay, maxDelay);
    return delay;
  }

  void _stopReconnectTimer() {
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
  }

  // Ping timer
  void _startPingTimer() {
    _stopPingTimer();
    _pingTimer = Timer.periodic(
      const Duration(seconds: 10),
      (_) => sendPing(),
    );
  }

  void _stopPingTimer() {
    _pingTimer?.cancel();
    _pingTimer = null;
  }

  // State management
  void _updateConnectionState(WebSocketConnectionState newState) {
    _connectionState = newState;
    notifyListeners();
  }

  // Initialize with automatic connection
  Future<void> initialize() async {
    // Try to connect to the default server
    await connect();
  }

  // Debug helpers
  String get debugInfo {
    return '''
WebSocketStore Debug Info:
- Status: ${_connectionState.status}
- URL: ${_connectionState.url}
- Reconnect Attempts: ${_connectionState.reconnectAttempts}/${_connectionState.maxReconnectAttempts}
- Last Ping: ${_connectionState.lastPing}
- Last Send: ${_connectionState.lastSend}
- Last Receive: ${_connectionState.lastReceive}
- Selected Server: $_selectedServerId
- Current Server Index: $_currentServerIndex
- Error: ${_connectionState.errorMessage ?? 'None'}
''';
  }

  @override
  void dispose() {
    _stopReconnectTimer();
    _stopPingTimer();
    disconnect();
    super.dispose();
  }
} 